src/core/DY.Core.Trace.SplitUtils.fst

```diff
@@ -0,0 +1,273 @@
+    )
+  = norm_spec [zeta; delta_only [`%trace_split]] (trace_split #a);
+    norm_spec [zeta; delta_only [`%trace_split]] (trace_split #b);
+    let Snoc hd entry = tr in
+    if ts = last_ts tr then ()
+    else trace_split_fmap hd ts f
+
+(*** Trace arithmetic ***)
+
+/// We define trace concatenation (the monoid operation), as well as
+/// trace subtraction, which is defined for traces tr2 <$ tr1, and yields the suffix of
+/// tr1 beginning after tr2.
+
+[@@"opaque_to_smt"]
+let rec trace_concat (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : trace_ label_t
+  = match tr2 with
+    | Nil -> tr1
+    | Snoc hd e -> Snoc (trace_concat tr1 hd) e
+
+[@@"opaque_to_smt"]
+let rec trace_subtract (#label_t:Type) (tr1:trace_ label_t) (tr2:trace_ label_t{tr2 <$ tr1})
+  : trace_ label_t
+  = if length tr1 = length tr2 then Nil
+    else begin
+      let Snoc hd e = tr1 in
+      grows_cases tr2 tr1;
+      Snoc (trace_subtract hd tr2) e
+    end
+
+let (<++>) = trace_concat
+let (<-->) = trace_subtract
+
+/// Traces form a monoid with trace concatenation as operation, and Nil as unit.
+
+let rec trace_concat_nil (#label_t:Type) (tr:trace_ label_t)
+  : Lemma
+    (
+      tr <++> Nil == tr /\
+      Nil <++> tr == tr
+    )
+    [SMTPatOr [[SMTPat (tr <++> Nil)]; [SMTPat (Nil <++> tr)]]]
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t);
+    match tr with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_nil hd
+
+let rec trace_concat_assoc (#label_t:Type) (tr1 tr2 tr3:trace_ label_t)
+  : Lemma
+    (((tr1 <++> tr2) <++> tr3) == (tr1 <++> (tr2 <++> tr3)))
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t);
+    match tr3 with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_assoc tr1 tr2 hd
+
+let snoc_is_concat_singleton (#label_t:Type) (tr :trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (Snoc tr e == (tr <++> (Snoc Nil e)))
+    [SMTPat (tr <++> (Snoc Nil e))]
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t)
+
+let rec trace_concat_grows (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (tr1 <$ (tr1 <++> tr2))
+    [SMTPat (tr1 <$ (tr1 <++> tr2))]
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t);
+    match tr2 with
+    | Nil -> ()
+    | Snoc hd e -> trace_concat_grows tr1 hd
+
+/// Properties of trace subtraction
+
+let trace_subtract_snoc_left (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures ((Snoc tr2 e) <--> tr1) == Snoc (tr2 <--> tr1) e)
+  = norm_spec [zeta; delta_only [`%trace_subtract]] (trace_subtract #label_t)
+
+let trace_subtract_snoc_left' (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires Snoc? tr2 /\ (let Snoc hd _ = tr2 in tr1 <$ hd))
+    (ensures (
+      let Snoc hd e = tr2 in
+      (tr2 <--> tr1) == Snoc (hd <--> tr1) e
+    ))
+  = norm_spec [zeta; delta_only [`%trace_subtract]] (trace_subtract #label_t)
```

Comment:
The second of these lemmas seems more often practically useful, while the first is much simpler to write. Maybe there's either a more fundamental way to get at the same property, or a way to give an SMTPat for one or the other that avoids the need to manually invoke them.

For instance, `[SMTPat (Snoc tr2 e) <--> tr1]` as an SMTPat on `trace_subtract_snoc_left` makes `trace_subtract_snoc_left'` prove automatically, and also allows me to remove any explicit calls to `trace_subtract_snoc_left'` in the remaining lemmas. I think this direction is the one we want most of the time, but am not sure if this SMTPat is ideal.

Reply:
@TWal maybe you have more insight on this? It looks fine for me so far, but I'm not sure if this SMTPat will trigger too often --- in particular, I'm not sure if `tr2 <--> tr1` and `Snoc? tr2` together will cause this to trigger, which would mean it tries to trigger every time we do any nontrivial subtraction, or only if we directly introduce `Snoc _ _ <--> _` into the context.

Reply:
Ah sorry, I missed your question.

I had no intuition on whether `Snoc? tr` would trigger the lemma, so I made the following experiment that seems to say that it does trigger (if ifuel is >= 1):

```fstar
assume val p: list int -> prop
assume val lem:
  h:int -> t:list int ->
  Lemma
  (ensures p (h::t))
  [SMTPat (p (h::t))]

val lem2:
  l:list int ->
  Lemma
  (requires Cons? l)
  (ensures p l)
let lem2 l =
  () // proof is automatic thanks to SMTPat above!
```

I however think that if we use `append_entry` from #92 then it will be fine, as shown by the following experiment:
```fstar
let append_entry (h:int) (t:list int) = h::t

assume val p: list int -> prop
assume val lem:
  h:int -> t:list int ->
  Lemma
  (ensures p (append_entry h t))
  [SMTPat (p (append_entry h t))]

val lem2:
  l:list int ->
  Lemma
  (requires Cons? l)
  (ensures p l)
let lem2 l =
  () // now fails
```

Reply:
This seemed to work quite well, thanks.